devito/petsc/solve.py

```diff
@@ -105,3 +115,51 @@ def _(expr, target):
 @remove_target.register(Derivative)
 def _(expr, target):
     return 0 if expr.has(target) else expr
+
+
+def centre_stencil(eqn, target):
+    """
+    Extract the centre stencil from equation.
+    This function assumes that the core stencil, which is the stencil from
+    which the centre stencil will be extracted, is located on the RHS of
+    the input equation (first argument).
+
+    NOTE: At the point of entry, the time derivatives are likely evaluated, but
+    not the spatial derivatives. This necessitates evaluating 'eqn'
```

Comment:
you cannot pre-evaluate an expression -- thus destroying its (e.g.) `IndexDerivative` objects -- to obtain this information... assuming I'm reading this docstring correctly. You'd be hurting compilation elsewhere. 

Can you add an example of what you're trying to accomplish here? Also, what's the use case?

Reply:
So looks like this only locally evaluate it to extract the middle part so it shouldn't impact the rest of the  compiler.
evaluate can be a little bit expensive though might wanna add a TODO that needs to be improved somehow